go/ct/spc/specification.go

```diff
@@ -2256,48 +2262,72 @@ func nonStaticSelfDestructRules(revision tosca.Revision, warm bool, destinationC
 			Eq(ReadOnly(), false),
 			IsRevision(revision),
 			hasSelfDestructedCondition,
-			targetWarm,
+			beneficiaryExists,
+			beneficiaryWarm,
 		},
 		parameters: []Parameter{AddressParameter{}},
 		effect: func(s *st.State) {
-			selfDestructEffect(s, destinationColdCost, accountCreationFee, refundGas)
+			selfDestructEffect(s)
 		},
 	}
 
 	return rulesFor(instruction)
 }
 
-func selfDestructEffect(s *st.State, destinationColdCost, accountCreationFee, refundGas tosca.Gas) {
+func selfDestructEffect(s *st.State) {
 	// Behavior pre cancun: the current account is registered to be destroyed, and will be at the end of the current
 	// transaction. The transfer of the current balance to the given account cannot fail. In particular,
 	// the destination account code (if any) is not executed, or, if the account does not exist, the
 	// balance is still added to the given address.
 
-	// account to send the current balance to
-	destinationAccount := s.Stack.Pop().Bytes20be()
-	currentAccount := s.CallContext.AccountAddress
-	CurrentBalance := s.Accounts.GetBalance(currentAccount)
+	beneficiaryAccount := s.Stack.Pop().Bytes20be()
+	originatorAccount := s.CallContext.AccountAddress
+	originatorBalance := s.Accounts.GetBalance(originatorAccount)
 
 	dynamicCost := tosca.Gas(0)
-	if !CurrentBalance.IsZero() && !s.Accounts.Exist(destinationAccount) {
-		dynamicCost += accountCreationFee
-	}
 
-	dynamicCost += destinationColdCost
+	// Add costs for transfering the remaining balance.
+	if !originatorBalance.IsZero() {
+		// If the target account does not exist, the account creation fee is added.
+		if !s.Accounts.Exists(beneficiaryAccount) {
```

Comment:
evm codes says: "State: is the beneficiary account empty?"  +25k if yes
" If a positive balance is sent to an empty account, the dynamic gas is 25000" 

The way I understand it, all non-existing accounts are empty, but not otherwise.

Reply:
python reference: 
```python
  if (
        not is_account_alive(evm.env.state, beneficiary)
        and get_account(evm.env.state, evm.message.current_target).balance != 0
    ):
        gas_cost += GAS_SELF_DESTRUCT_NEW_ACCOUNT
```

where is_account_alive is
```python
   account = get_account_optional(state, address)
    if account is None:
        return False
    else:
        return not (
            account.nonce == Uint(0)
            and account.code == b""
            and account.balance == 0
        )
```

```

Reply:
I really think this should be "empty" and not just "exists"

Reply:
Right, thanks for checking. Also geth uses the `Empty` check [here](https://github.com/ethereum/go-ethereum/blob/3ff73d46b3592ca37175021b9aa3cb5187698070/core/vm/gas_table.go#L492).